File: src-1.x/BYTES.NET/WPF/MVVM/GUIThreadViewModel.vb
'import .net namespace(s) required
Imports System.Threading
Imports System.Windows.Threading

Namespace WPF.MVVM

    Public MustInherit Class GUIThreadViewModel(Of T)

        Inherits ViewModel

#Region "public event(s)"

        Public Event Closed(sender As GUIThreadViewModel(Of T))

#End Region

#Region "protected variable(s)"

        Protected _myView As Window = Nothing
        Protected _myThread As Thread = Nothing

#End Region

#Region "public properties"

        Public ReadOnly Property View As Window
            Get

                Return _myView

            End Get
        End Property

        Public ReadOnly Property ThreadID As Integer
            Get

                Return _myThread.ManagedThreadId

            End Get
        End Property

#End Region

#Region "public new instance method(s)"

        ''' <summary>
        ''' default new instance method
        ''' </summary>
        ''' <param name="show"></param>
        Public Sub New(Optional ByVal show As Boolean = True)

            'create a new base-class instance
            MyBase.New()

            'set the variable(s)
            If show Then

                Me.Show()

            End If

        End Sub

#End Region

#Region "public method(s)"

        ''' <summary>
        ''' method showing up the GUI
        ''' </summary>
        Public Sub Show()

            'clear from a possibly running instance
            If Not IsNothing(_myThread) Then

                RaiseEvent Closed(Me)

                _myThread.Abort()
                _myThread = Nothing

            End If

            'start a new thread
            _myThread = New Thread(AddressOf ShowWindow)
            _myThread.SetApartmentState(ApartmentState.STA)
            _myThread.Start()

        End Sub

        ''' <summary>
        ''' method terminating the thread
        ''' </summary>
        Public Sub Close()

            If Not IsNothing(_myThread) Then

                RaiseEvent Closed(Me)

                _myThread.Abort()

            End If

        End Sub

#End Region

#Region "private method(s)"

        ''' <summary>
        ''' method showing up the GUI initially
        ''' </summary>
        Private Sub ShowWindow()

            'create a new view instance
            _myView = DirectCast(Activator.CreateInstance(GetType(T)), Window)
            _myView.DataContext = Me
            _myView.Show()

            AddHandler _myView.Closed, AddressOf OnWindowClosed

            Dispatcher.Run()

        End Sub

        ''' <summary>
        ''' event handler for method on GUI window closed 
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        Private Sub OnWindowClosed(sender, e)

            RaiseEvent Closed(Me)

            _myView.Dispatcher.InvokeShutdown()

        End Sub

#End Region

    End Class

End Namespace

File: src-2023/BYTES.NET.WPF/MVVM/GUIThreadViewModel.cs
using System;
using System.Threading;
using System.Windows;
using System.Windows.Threading;

namespace BYTES.NET.WPF.MVVM
{
    /// <summary>
    /// ViewModel for a GUI Thread
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public abstract class GUIThreadViewModel<T> : ViewModel where T : Window, new()
    {
        /// <summary>
        /// View for the ViewModel
        /// </summary>
        protected Window _myView;

        /// <summary>
        /// Thread for the ViewModel
        /// </summary>
        protected Thread _myThread;

        /// <summary>
        /// Cancellation token source for the ViewModel
        /// </summary>
        // went with cancellation token refer to https://learn.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads
        private CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();

        /// <summary>
        /// Lambda expression for the View
        /// </summary>
        public Window View => _myView;

        /// <summary>
        /// Thread ID
        /// </summary>
        public int ThreadID => _myThread.ManagedThreadId;

        public event Action<GUIThreadViewModel<T>> Closed;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="show"></param>
        public GUIThreadViewModel(bool show = true)
        {
            if (show)
            {
                Show();
            }
        }

        /// <summary>
        /// Shows the ViewModel
        /// </summary>
        public void Show()
        {
            if (_myThread != null)
            {
                this.Closed?.Invoke(this);
                _cancellationTokenSource.Cancel(); // Cancel the previous token
                _myThread.Join(); // Wait for the previous thread to finish
                _myThread = null;
            }
            _cancellationTokenSource = new CancellationTokenSource(); // Create a new token source
            _myThread = new Thread(() => ShowWindow(_cancellationTokenSource.Token));
            _myThread.SetApartmentState(ApartmentState.STA);
            _myThread.Start();
        }

        /// <summary>
        /// Closes the ViewModel
        /// </summary>
        public void Close()
        {
            if (_myThread != null)
            {
                _cancellationTokenSource.Cancel(); // Cancel the token
                _myThread.Join(); // Wait for the thread to finish
                _myThread = null;
            }
        }

        private void ShowWindow(CancellationToken cancellationToken)
        {
            _myView = new T();
            _myView.DataContext = this;
            _myView.Show();
            _myView.Closed += OnWindowClosed;
            Dispatcher.Run();
        }

        private void OnWindowClosed(object sender, EventArgs e)
        {
            this.Closed?.Invoke(this);
            _myView.Dispatcher.InvokeShutdown();
        }
    }
}
